### PR TITLE
Implement clearcoat per the Filament and the `KHR_materials_clearcoat` specifications.

### PR DESCRIPTION
**Pull request by pcwalton**
Clearcoat is a separate material layer that represents a thin translucent layer of a material. Examples include (from the [Filament spec]) car paint, soda cans, and lacquered wood. This commit implements support for clearcoat following the Filament and Khronos specifications, marking the beginnings of support for multiple PBR layers in Bevy.

The [`KHR_materials_clearcoat`] specification describes the clearcoat support in glTF. In Blender, applying a clearcoat to the Principled BSDF node causes the clearcoat settings to be exported via this extension. As of this commit, Bevy parses and reads the extension data when present in glTF. Note that the `gltf` crate has no support for `KHR_materials_clearcoat`; this patch therefore implements the JSON semantics manually.

Clearcoat is integrated with `StandardMaterial`, but the code is behind a series of `#ifdef`s that only activate when clearcoat is present. Additionally, the `pbr_feature_layer_material_textures` Cargo feature must be active in order to enable support for clearcoat factor maps, clearcoat roughness maps, and clearcoat normal maps. This approach mirrors the same pattern used by the existing transmission feature and exists to avoid running out of texture bindings on platforms like WebGL and WebGPU. Note that constant clearcoat factors and roughness values *are* supported in the browser; only the relatively-less-common maps are disabled on those platforms.

This patch refactors the lighting code in `StandardMaterial` significantly in order to better support multiple layers in a natural way. That code was due for a refactor in any case, so this is a nice improvement.

A new demo, `clearcoat`, has been added. It's based on [the corresponding three.js demo], but all the assets (aside from the skybox and environment map) are my original work.

[Filament spec]: https://google.github.io/filament/Filament.html#materialsystem/clearcoatmodel

[`KHR_materials_clearcoat`]: https://github.com/KhronosGroup/glTF/blob/main/extensions/2.0/Khronos/KHR_materials_clearcoat/README.md

[the corresponding three.js demo]: https://threejs.org/examples/webgl_materials_physical_clearcoat.html

Screenshot 2024-04-19 101143

Screenshot 2024-04-19 102054

### Added

* `StandardMaterial` now supports a clearcoat layer, which represents a thin translucent layer over an underlying material.
* The glTF loader now supports the `KHR_materials_clearcoat` extension, representing materials with clearcoat layers.

## Migration Guide

* The lighting functions in the `pbr_lighting` WGSL module now have clearcoat parameters, if `STANDARD_MATERIAL_CLEARCOAT` is defined.

* The `R` reflection vector parameter has been removed from some lighting functions, as it was unused.

### CODE REVIEW

File: crates/bevy_gltf/src/loader.rs
Comment (robtfm) [inline crates/bevy_gltf/src/loader.rs:1713]: could this be upstreamed?

File: crates/bevy_pbr/src/render/pbr_fragment.wgsl
Comment (robtfm) [inline crates/bevy_pbr/src/render/pbr_fragment.wgsl:182]: variants of this block are repeated 10 times now, could we factor it out?

File: crates/bevy_pbr/src/render/pbr_functions.wgsl
Comment (robtfm) [inline crates/bevy_pbr/src/render/pbr_functions.wgsl:72]: just calling out for other reviewers - this function is no longer called unless VERTEX_UVS / VERTEX_TANGENTS are active so the behaviour changes in the non-tangent case don't matter.

but also @pcwalton - array_texture.wgsl needs updating for this api change.

File: crates/bevy_pbr/src/render/pbr_functions.wgsl
Comment (robtfm) [inline crates/bevy_pbr/src/render/pbr_functions.wgsl:244]: this doesn't look right - you #ifdef'd 5 params in the point_light call above.

File: crates/bevy_pbr/src/render/pbr_functions.wgsl
Comment (robtfm) [inline crates/bevy_pbr/src/render/pbr_functions.wgsl:225]: maybe we could have a struct to hold all these arguments ... or most of them, so we could reuse it with a couple extra params for all the lighting calls?

File: crates/bevy_pbr/src/pbr_material.rs
Comment (robtfm) [inline crates/bevy_pbr/src/pbr_material.rs:829]: nitpick - these flags should probably be added in the prepass as well in case anybody wants to use them

**Review by robtfm — COMMENTED** [review]
just a first pass. the changes to the core lighting functions are pretty extensive, @mockersf could we please run a demo comparison? 

i think we need to make proper efforts to stop the lighting code becoming more write-only than it already is. not saying this makes it much worse than it already was, but it's honestly very hard to follow.

it would also be nice to improve the example a bit (the three.js one is nicer, and i don't think their implementation is better) but that can certainly be done later.

**Comment by pcwalton** [comment]
@robtfm How would you like me to improve the example? I wanted to include a transparent one because @JMS55 asked me how clearcoat looked with transparency. Should I add more of a noise pattern to the car paint? Redo the golf ball texture?

File: crates/bevy_gltf/src/loader.rs
Comment (pcwalton) [reply crates/bevy_gltf/src/loader.rs:1713]: I can, but `gltf` hasn't had any updates for a month. Do you want to block this PR on that?

**Comment by pcwalton** [comment]
@robtfm I tried to improve the lighting code as I went and make it as easy to follow as possible. Sorry to hear that I didn't succeed :( Do you have any suggestions?

File: crates/bevy_pbr/src/render/pbr_functions.wgsl
Comment (pcwalton) [reply crates/bevy_pbr/src/render/pbr_functions.wgsl:72]: Good catch, fixed.

File: crates/bevy_pbr/src/render/pbr_functions.wgsl
Comment (pcwalton) [reply crates/bevy_pbr/src/render/pbr_functions.wgsl:225]: Working on this. This is a huge change :)

File: crates/bevy_pbr/src/render/pbr_fragment.wgsl
Comment (pcwalton) [reply crates/bevy_pbr/src/render/pbr_fragment.wgsl]: Done.

**Comment by robtfm** [comment]
> @robtfm How would you like me to improve the example? I wanted to include a transparent one because @JMS55 asked me how clearcoat looked with transparency. Should I add more of a noise pattern to the car paint? Redo the golf ball texture?

i think the main issue i have is the specular aliasing from the normal maps. it seems less "bitty" in the 3js version. perhaps just downres the normal map, or just use a smaller one maybe so the single pixel highlights are less frequent?
the texture on the carpaint is nicer in 3js as well (and the spinning spheres) but i'm not sure that's worth adding a new texture for.



> @robtfm I tried to improve the lighting code as I went and make it as easy to follow as possible. Sorry to hear that I didn't succeed :( Do you have any suggestions?

yes sorry, i didn't mean to sound so critical. i think if you hadn't made such an effort it would be much worse, whereas now it's not much worse, and with more functionality. the 2 specific comments about a param struct and factoring the meshlet sample were the main things i thought would help. i guess if the lighting functions api is clean then at least the trickier stuff is more contained.

File: crates/bevy_pbr/src/render/pbr_functions.wgsl
Comment (pcwalton) [reply crates/bevy_pbr/src/render/pbr_functions.wgsl]: This is now done. The entire `pbr_lighting.wgsl` file has been refactored from top to bottom to use structures throughout. I also changed `environment_map.wgsl` to do the same

File: crates/bevy_pbr/src/pbr_material.rs
Comment (pcwalton) [reply crates/bevy_pbr/src/pbr_material.rs:829]: I don't know what this means. The prepass doesn't care about any of these flags, and we don't do this anywhere else (grepping for e.g. `BASE_COLOR` brings up nothing related to prepasses, except for alpha discard, which isn't relevant here).

**Comment by pcwalton** [comment]
@robtfm I believe every comment has been addressed. `pbr_lighting.wgsl` has been largely rewritten. I made the spheres rotate, added some blur and decreased the levels on the scratch normal map, added some subtle blue noise as a normal map to the car paint, and switched to ACES filmic tonemapping, as this is what makes the normal map on the car paint in three.js appear purple.

Screenshot 2024-04-22 160935

**Comment by mockersf** [comment]
> just a first pass. the changes to the core lighting functions are pretty extensive, @mockersf could we please run a demo comparison?

Mostly good, just a change in the transmission example:

**Comment by pcwalton** [comment]
How do I get rid of the GitHub typos check on `Fo`?

**Comment by mockersf** [comment]
add it to https://github.com/bevyengine/bevy/blob/main/typos.toml

**Comment by pcwalton** [comment]
I *think* I fixed the transmission example, but it's hard for me to check locally as it's animated.

File: crates/bevy_pbr/src/light_probe/environment_map.wgsl
Comment (JMS55) [inline crates/bevy_pbr/src/light_probe/environment_map.wgsl:182]: What bug? Please open an issue with wgpu and link to it here.

**Comment by robtfm** [comment]
the gold and golf balls looks great now, but the blue sphere doesn't show: `ERROR bevy_asset::server: Path not found: [...]/assets/textures/BlueNoise-Normal.png`

**Comment by pcwalton** [comment]
@robtfm Oops, forgot to check that file in! Fixed.

File: crates/bevy_pbr/src/light_probe/environment_map.wgsl
Comment (pcwalton) [reply crates/bevy_pbr/src/light_probe/environment_map.wgsl]: https://github.com/gfx-rs/wgpu/issues/5593, added a link

File: crates/bevy_gltf/Cargo.toml
Comment (JMS55) [inline crates/bevy_gltf/Cargo.toml:14]: I feel like we should split StandardMaterial into 2 / features / something. A basic one with the typical PBR stuff for lower end platforms, and one with all features included for newer platforms. Doesn't have to be this PR.

File: crates/bevy_pbr/src/light_probe/environment_map.wgsl
Comment (JMS55) [inline crates/bevy_pbr/src/light_probe/environment_map.wgsl:175]: This file is getting rough 😅 . I won't block on it, but I'd prefer to start factoring it out in multiple functions with this file/other files.

File: crates/bevy_pbr/src/pbr_material.rs
Comment (JMS55) [inline crates/bevy_pbr/src/pbr_material.rs:342]: I would remove the mention of filament.

File: crates/bevy_pbr/src/pbr_material.rs
Comment (JMS55) [inline crates/bevy_pbr/src/pbr_material.rs:371]:
```suggestion
    /// the green channel. Values from this texture are multiplied by the main
```

File: crates/bevy_pbr/src/pbr_material.rs
Comment (JMS55) [reply crates/bevy_pbr/src/pbr_material.rs:829]: Yeah I wouldn't add them. Sometime soon we should revisit how pipeline keys are managed however.

File: crates/bevy_pbr/src/render/pbr_bindings.wgsl
Comment (JMS55) [inline crates/bevy_pbr/src/render/pbr_bindings.wgsl:32]: The continued per-texture sampler is rough, but I don't think there's a better solution until we have bindless and can use indices into a small array of samplers.

File: crates/bevy_pbr/src/render/pbr_types.wgsl
Comment (JMS55) [inline crates/bevy_pbr/src/render/pbr_types.wgsl:18]: You need to update standard_material_new() in this file too.

File: crates/bevy_gltf/Cargo.toml
Comment (pcwalton) [reply crates/bevy_gltf/Cargo.toml:14]: Yeah, I considered doing that here, but decided against it because of the precedent set by transmission, which is also a more advanced feature. My personal feeling is that maybe `StandardMaterial` should be the glTF standard with no extensions, and then we can have `GltfExtensionMaterial`, `OpenPbrMaterial`, `MaterialXMaterial` that build on top of it. But I imagine there will be a lot of discussion around this topic, so that's for a followup :)

File: crates/bevy_pbr/src/pbr_material.rs
Comment (pcwalton) [reply crates/bevy_pbr/src/pbr_material.rs:342]: Yeah, I just mentioned Filament there because I lifted the line from its documentation and was worried about plagiarism. I switched it to just "This is typically used for painted surfaces."

File: crates/bevy_pbr/src/render/pbr_bindings.wgsl
Comment (pcwalton) [reply crates/bevy_pbr/src/render/pbr_bindings.wgsl:32]: Yeah, I really don't like it either. (In general glTF may have gone overboard with all the textures at the standards level; if you're really getting that exact about e.g. clearcoat roughness you probably want a node graph like MaterialX anyway. But standards are standards, so I added support anyway...)

File: crates/bevy_pbr/src/light_probe/environment_map.wgsl
Comment (pcwalton) [reply crates/bevy_pbr/src/light_probe/environment_map.wgsl:175]: It's only 198 lines, which is pretty short. How were you thinking about splitting it up?

Comment (JMS55) [reply crates/bevy_pbr/src/light_probe/environment_map.wgsl:175]: Splitting the clear coat stuff up to a separate function would help maybe. You're right it's not so bad though (I was looking at it in the diff view).

File: crates/bevy_pbr/src/render/pbr_fragment.wgsl
Comment (robtfm) [inline crates/bevy_pbr/src/render/pbr_fragment.wgsl:88]: perhaps it would be better if `sample_texture` took a struct who's declaration varied based on the MESHLET define?

not a big deal though

File: crates/bevy_pbr/src/light_probe/environment_map.wgsl
Comment (robtfm) [inline crates/bevy_pbr/src/light_probe/environment_map.wgsl:184]: from the duped issue this doesn't sound like a bug, just an unsupported extension (`unrestricted_pointer_parameters`).

not clear to me if this is an extension we can request / something that's supported in some contexts though.

File: crates/bevy_pbr/src/render/pbr_lighting.wgsl
Comment (robtfm) [inline crates/bevy_pbr/src/render/pbr_lighting.wgsl:45]: we might benefit from reordering fields here - i think the f32 ahead of the 16b aligned vec3s wastes some space. these are not i/o structs so i don't know how important it is.

**Review by robtfm — COMMENTED** [review]
i think it's a good improvement, though passing by reference really needs some ergonomics improvements (why no `->` wgsl?!).

File: crates/bevy_pbr/src/render/pbr_lighting.wgsl
Comment (JMS55) [reply crates/bevy_pbr/src/render/pbr_lighting.wgsl:197]: It's in the webgpu spec now, just not implemented by naga yet. "pointer composite access" https://gpuweb.github.io/gpuweb/wgsl/#composite-value-decomposition-expr

File: crates/bevy_pbr/src/light_probe/environment_map.wgsl
Comment (robtfm) [reply crates/bevy_pbr/src/light_probe/environment_map.wgsl:184]: Could we make this work by passing around an actual struct that contains pointers to a base, a base_layer and a clearcoat_layer?

File: crates/bevy_pbr/src/render/pbr_functions.wgsl
Comment (pcwalton) [reply crates/bevy_pbr/src/render/pbr_functions.wgsl:218]: I would guess that any shader compiler worth its salt is going to inline and get rid of the struct anyway.

File: crates/bevy_pbr/src/light_probe/environment_map.wgsl
Comment (pcwalton) [reply crates/bevy_pbr/src/light_probe/environment_map.wgsl:184]: You can't put `ptr<function, ...>` types in structs in WGSL. I just tried.

Comment (robtfm) [reply crates/bevy_pbr/src/light_probe/environment_map.wgsl:184]: Right, that is explicitly denied in the spec. Maybe use a fixed-size array of layers and pass the full ptr and an index into the layer-based functions..?

I don’t know if this would be worthwhile; I can try it and profile later if you don’t want to (I realise I’ve been pretty demanding already)

File: crates/bevy_pbr/src/render/pbr_functions.wgsl
Comment (robtfm) [reply crates/bevy_pbr/src/render/pbr_functions.wgsl:459]: i tried to get this to fail, but couldn't - finally realized it's because `STANDARD_MATERIAL_SPECULAR_OR_DIFFUSE_TRANSMISSION` is never defined.

File: crates/bevy_pbr/src/render/pbr_functions.wgsl
Comment (pcwalton) [reply crates/bevy_pbr/src/render/pbr_functions.wgsl]: I checked the RDNA disassembly and it is inlined and the structs don't exist at runtime. I haven't checked NVIDIA and Intel but I suspect it's the same. I'm going to go ahead and resolve this.

File: crates/bevy_pbr/src/render/pbr_fragment.wgsl
Comment (pcwalton) [reply crates/bevy_pbr/src/render/pbr_fragment.wgsl]: Done.

File: crates/bevy_pbr/src/light_probe/environment_map.wgsl
Comment (pcwalton) [reply crates/bevy_pbr/src/light_probe/environment_map.wgsl]: OK, I did that (layers and indices).

File: crates/bevy_pbr/src/render/pbr_lighting.wgsl
Comment (pcwalton) [reply crates/bevy_pbr/src/render/pbr_lighting.wgsl:48]: Reordered.

File: crates/bevy_pbr/src/render/pbr_functions.wgsl
Comment (pcwalton) [reply crates/bevy_pbr/src/render/pbr_functions.wgsl]: Fixed.

File: crates/bevy_pbr/src/light_probe/environment_map.wgsl
Comment (pcwalton) [reply crates/bevy_pbr/src/light_probe/environment_map.wgsl:202]: OK, done.

**Review by robtfm — APPROVED** [review]
looks good!

we should probably add something to the migration guide mentioning the changes required for custom prepasses/shaders that use `apply_normal_mapping`.

**Comment by superdump** [comment]
I’ve read the discussion, has any GPU-profiling/benchmarking been done across vendors to see what impact the shader refactoring has on performance without clear coat?

**Comment by pcwalton** [comment]
On NVIDIA, warp occupancy and register count are identical both before and after this PR.

Screenshot 2024-05-01 150843

Screenshot 2024-05-01 151432

**Comment by NthTensor** [comment]
No obvious performance differences for me testing vs main. What's blocking this moving forward? The more general changes seem like a marked improvement in standardization and code quality.

**Comment by pcwalton** [comment]
@NthTensor It only has one approval. Having a second approval would be very helpful.

**Review by JMS55 — APPROVED** [review]

**Comment by pcwalton** [comment]
Closing as I am burned out and have no motivation to get this landed anymore.

**Comment by alice-i-cecile** [comment]
@pcwalton once merge conflicts are resolved this should be good to merge. Do reach out if you'd like a hand with that.

**Comment by pcwalton** [comment]
Updated.

**Comment by alice-i-cecile** [comment]
I'm comfortable with the thoroughness and expertise of the reviews here, and this isn't a controversial addition. Merging.

**Pull request merged**
